package/MDAnalysis/lib/util.py

```diff
@@ -2051,32 +2070,54 @@ def check_coords_decorator(func):
-                                                      coords.shape))
-                if coords.ndim == 1:
-                    is_single = True
-                    if convert_single:
-                        coords = coords[None, :]
+            if isinstance(coords, np.ndarray):
+                if allow_single:
+                    if (coords.ndim not in (1, 2)) or (coords.shape[-1] != 3):
+                        raise ValueError("{}(): {}.shape must be (3,) or"
+                                         " (n, 3), got {}.".format(fname,
+                                         argname, coords.shape))
+                    if coords.ndim == 1:
+                        is_single = True
+                        if convert_single:
+                            coords = coords[None, :]
+                else:
+                    if (coords.ndim != 2) or (coords.shape[1] != 3):
+                        raise ValueError("{}(): {}.shape must be (n, 3),"
+                                         " got {}.".format(fname, argname,
+                                                          coords.shape))
+                if enforce_dtype:
+                    try:
+                        coords = coords.astype(
+                            np.float32, order='C', copy=enforce_copy)
+                    except ValueError:
+                        errmsg = (f"{fname}(): {argname}.dtype must be"
+                                  f"convertible to float32, got"
+                                  f" {coords.dtype}.")
+                        raise TypeError(errmsg) from None
+                # coordinates should now be the right shape
+                ncoord = coords.shape[0]
             else:
-                if (coords.ndim != 2) or (coords.shape[1] != 3):
-                    raise ValueError("{}(): {}.shape must be (n, 3), got {}."
-                                     "".format(fname, argname, coords.shape))
-            if enforce_dtype:
                 try:
-                    coords = coords.astype(
-                        np.float32, order='C', copy=enforce_copy)
-                except ValueError:
-                    errmsg = (f"{fname}(): {argname}.dtype must be convertible to "
-                              f"float32, got {coords.dtype}.")
-                    raise TypeError(errmsg) from None
-            return coords, is_single
+                    coords = coords.positions  # homogenise to a numpy array
+                    ncoord = coords.shape[0]
+                    if allow_atomgroup:
+                        pass
+                    else:
+                        err = TypeError("AtomGroup or other class with a"
+                                        "`.positions` method supplied as an"
+                                        "argument, but allow_atomgroup is"
+                                        " False")
+                        raise err
+                except TypeError:
+                    raise err
```

Comment:
Shouldn't this except just be merged with the subsequent one? I don't quite understand the logic here.

Reply:
I'm sure it can be done better as it definitely looks janky but I think my logic was:

If an `AtomGroup` is supplied but not allowed by `allow_atomgroup` we need to manually raise an exception inside the `try `block. 

If we get this `TypeError` then re-raise it in the `except: TypeError` block. 

Otherwise we catch an `AttributeError` and raise a TypeError that the wrong type was supplied (see updated code).

Reply:
With the explanation it makes a bit more sense. It's just confusing (and a bit fragile) because you assume that the only place where a TypeError can originate is L2109 where you raise it yourself. Then you catch the exception to re-raise. 

Why do you have to re-raise? Can't you just delete
```python
except TypeError:
   raise err
```
and just `raise TypeError("AtomGroup or other class ...")`. The outer `except AttributeError` will not swallow the `TypeError`.

Perhaps you needed the re-raise when you had the blank `except:` but hopefully not anymore.